# Data Assignment 1 - `t54zheng` (20939203)


In [1]:
import pandas as pd
from pandasql import sqldf
import numpy as np

## Part 1 - Downloading and Reading Raw Data

WRDS Queries
- [8076602 - crsp](https://wrds-www.wharton.upenn.edu/query-manager/query/8076602/)
- [8076622 - compustat](https://wrds-www.wharton.upenn.edu/query-manager/query/8076622/)


In [2]:
# import raw data
crsp = pd.read_sas("crsp_nasdaq_100.sas7bdat", encoding = 'latin-1')
compustat = pd.read_sas("compustat_nasdaq_100.sas7bdat", encoding = 'latin-1')

In [3]:
crsp

PERMNO       DATE     CUSIP          COMNAM TICKER         PRC  \
0      10107.0 1997-01-31  59491810  MICROSOFT CORP   MSFT  102.000000   
1      10107.0 1997-02-28  59491810  MICROSOFT CORP   MSFT   97.500000   
2      10107.0 1997-03-31  59491810  MICROSOFT CORP   MSFT   91.687500   
3      10107.0 1997-04-30  59491810  MICROSOFT CORP   MSFT  121.500000   
4      10107.0 1997-05-30  59491810  MICROSOFT CORP   MSFT  124.000000   
...        ...        ...       ...             ...    ...         ...   
23992  93436.0 2022-08-31  88160R10       TESLA INC   TSLA  275.609985   
23993  93436.0 2022-09-30  88160R10       TESLA INC   TSLA  265.250000   
23994  93436.0 2022-10-31  88160R10       TESLA INC   TSLA  227.539993   
23995  93436.0 2022-11-30  88160R10       TESLA INC   TSLA  194.699997   
23996  93436.0 2022-12-30  88160R10       TESLA INC   TSLA  123.180000   

            RET      RETX     SHROUT    VWRETD    SPRTRN  
0      0.234493  0.234493  1198000.0  0.053473  0.061317  
1     -0.044118 -0.044118  1198000.0 -0.001067  0.005928  
2     -0.059615 -0.059615  1191000.0 -0.044889 -0.042614  
3      0.325153  0.325153  1191000.0  0.042396  0.058406  
4      0.020576  0.020576  1191000.0  0.071640  0.058577  
...         ...       ...        ...       ...       ...  
23992 -0.072489 -0.072489  3133470.0 -0.036233 -0.042440  
23993 -0.037589 -0.037589  3158000.0 -0.091323 -0.093396  
23994 -0.142168 -0.142168  3157752.0  0.077394  0.079863  
23995 -0.144326 -0.144326  3157752.0  0.052354  0.053753  
23996 -0.367334 -0.367334  3164000.0 -0.057119 -0.058971  

[23997 rows x 11 columns]

In [4]:
set(crsp.loc[crsp.PERMNO == 93436.0].TICKER)

{'TSLA', nan}

In [5]:
compustat

GVKEY   DATADATE  FYEARQ  FQTR                    CONM   TIC DATACQTR  \
0     001161 1996-03-31  1996.0   1.0  ADVANCED MICRO DEVICES   AMD   1996Q1   
1     001161 1996-06-30  1996.0   2.0  ADVANCED MICRO DEVICES   AMD   1996Q2   
2     001161 1996-09-30  1996.0   3.0  ADVANCED MICRO DEVICES   AMD   1996Q3   
3     001161 1996-12-31  1996.0   4.0  ADVANCED MICRO DEVICES   AMD   1996Q4   
4     001161 1997-03-31  1997.0   1.0  ADVANCED MICRO DEVICES   AMD   1997Q1   
...      ...        ...     ...   ...                     ...   ...      ...   
8858  184996 2021-12-31  2021.0   4.0               TESLA INC  TSLA   2021Q4   
8859  184996 2022-03-31  2022.0   1.0               TESLA INC  TSLA   2022Q1   
8860  184996 2022-06-30  2022.0   2.0               TESLA INC  TSLA   2022Q2   
8861  184996 2022-09-30  2022.0   3.0               TESLA INC  TSLA   2022Q3   
8862  184996 2022-12-31  2022.0   4.0               TESLA INC  TSLA   2022Q4   

     DATAFQTR        ATQ       CEQQ  EPSPXQ       IBQ      SALEQ CONSOL  \
0      1996Q1   2954.809   2121.948    0.18    25.327    544.212      C   
1      1996Q2   2845.025   2071.355   -0.26   -34.672    455.077      C   
2      1996Q3   3053.992   2036.658   -0.28   -38.362    456.862      C   
3      1996Q4   3145.283   2021.878   -0.15   -21.243    496.868      C   
4      1997Q1   3426.126   2043.757    0.09    12.951    551.999      C   
...       ...        ...        ...     ...       ...        ...    ...   
8858   2021Q4  62131.000  30189.000    2.29  2321.000  17719.000      C   
8859   2022Q1  66038.000  34085.000    3.20  3318.000  18756.000      C   
8860   2022Q2  68513.000  36376.000    2.18  2259.000  16934.000      C   
8861   2022Q3  74426.000  39851.000    1.05  3292.000  21454.000      C   
8862   2022Q4  82338.000  44704.000    1.18  3687.000  24318.000      C   

     INDFMT DATAFMT POPSRC CURCDQ COSTAT  
0      INDL     STD      D    USD      A  
1      INDL     STD      D    USD      A  
2      INDL     STD      D    USD      A  
3      INDL     STD      D    USD      A  
4      INDL     STD      D    USD      A  
...     ...     ...    ...    ...    ...  
8858   INDL     STD      D    USD      A  
8859   INDL     STD      D    USD      A  
8860   INDL     STD      D    USD      A  
8861   INDL     STD      D    USD      A  
8862   INDL     STD      D    USD      A  

[8863 rows x 19 columns]

### Data Cleanup

- Notice that there are more tickers in the crsp data than there are companies
- This means that some companies change ticker over the observed time period
- this means that we should actually include sort stocks by permno (as suggested in the assignment)

In [6]:
# Demonstrate this value is greater than 99 (size of our nasdaq 100 data)
print(f'CRSP Tickers - {len(set(crsp.TICKER))}')
print(f'Compustat Tickers - {len(set(compustat.TIC))}')

CRSP Tickers - 119
Compustat Tickers - 99


In [7]:
# To fix this, we will just add permnos to our compustat data
translation_df = pd.read_excel("nasdaq100_pop.xlsx", sheet_name="all", converters={"GVKEY":str, "PERMNO":str})
translation_dict = translation_df[["GVKEY", "PERMNO"]].set_index("GVKEY").to_dict()["PERMNO"]

In [8]:
compustat['PERMNO'] = compustat.apply(lambda row: translation_dict[row.GVKEY], axis=1)
compustat

GVKEY   DATADATE  FYEARQ  FQTR                    CONM   TIC DATACQTR  \
0     001161 1996-03-31  1996.0   1.0  ADVANCED MICRO DEVICES   AMD   1996Q1   
1     001161 1996-06-30  1996.0   2.0  ADVANCED MICRO DEVICES   AMD   1996Q2   
2     001161 1996-09-30  1996.0   3.0  ADVANCED MICRO DEVICES   AMD   1996Q3   
3     001161 1996-12-31  1996.0   4.0  ADVANCED MICRO DEVICES   AMD   1996Q4   
4     001161 1997-03-31  1997.0   1.0  ADVANCED MICRO DEVICES   AMD   1997Q1   
...      ...        ...     ...   ...                     ...   ...      ...   
8858  184996 2021-12-31  2021.0   4.0               TESLA INC  TSLA   2021Q4   
8859  184996 2022-03-31  2022.0   1.0               TESLA INC  TSLA   2022Q1   
8860  184996 2022-06-30  2022.0   2.0               TESLA INC  TSLA   2022Q2   
8861  184996 2022-09-30  2022.0   3.0               TESLA INC  TSLA   2022Q3   
8862  184996 2022-12-31  2022.0   4.0               TESLA INC  TSLA   2022Q4   

     DATAFQTR        ATQ       CEQQ  EPSPXQ       IBQ      SALEQ CONSOL  \
0      1996Q1   2954.809   2121.948    0.18    25.327    544.212      C   
1      1996Q2   2845.025   2071.355   -0.26   -34.672    455.077      C   
2      1996Q3   3053.992   2036.658   -0.28   -38.362    456.862      C   
3      1996Q4   3145.283   2021.878   -0.15   -21.243    496.868      C   
4      1997Q1   3426.126   2043.757    0.09    12.951    551.999      C   
...       ...        ...        ...     ...       ...        ...    ...   
8858   2021Q4  62131.000  30189.000    2.29  2321.000  17719.000      C   
8859   2022Q1  66038.000  34085.000    3.20  3318.000  18756.000      C   
8860   2022Q2  68513.000  36376.000    2.18  2259.000  16934.000      C   
8861   2022Q3  74426.000  39851.000    1.05  3292.000  21454.000      C   
8862   2022Q4  82338.000  44704.000    1.18  3687.000  24318.000      C   

     INDFMT DATAFMT POPSRC CURCDQ COSTAT PERMNO  
0      INDL     STD      D    USD      A  61241  
1      INDL     STD      D    USD      A  61241  
2      INDL     STD      D    USD      A  61241  
3      INDL     STD      D    USD      A  61241  
4      INDL     STD      D    USD      A  61241  
...     ...     ...    ...    ...    ...    ...  
8858   INDL     STD      D    USD      A  93436  
8859   INDL     STD      D    USD      A  93436  
8860   INDL     STD      D    USD      A  93436  
8861   INDL     STD      D    USD      A  93436  
8862   INDL     STD      D    USD      A  93436  

[8863 rows x 20 columns]

## Part 2 - Merging the data

In [9]:
# Enumerate data using monthid
crsp['YEAR'] = pd.DatetimeIndex(crsp['DATE']).year
crsp['MONTH'] = pd.DatetimeIndex(crsp['DATE']).month
crsp['monthid'] = ((crsp['YEAR']) - 1996) * 12 + crsp['MONTH']
sqldf("SELECT * FROM crsp LIMIT 5")

PERMNO                        DATE     CUSIP          COMNAM TICKER  \
0  10107.0  1997-01-31 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
1  10107.0  1997-02-28 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
2  10107.0  1997-03-31 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
3  10107.0  1997-04-30 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
4  10107.0  1997-05-30 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   

        PRC       RET      RETX     SHROUT    VWRETD    SPRTRN  YEAR  MONTH  \
0  102.0000  0.234493  0.234493  1198000.0  0.053473  0.061317  1997      1   
1   97.5000 -0.044118 -0.044118  1198000.0 -0.001067  0.005928  1997      2   
2   91.6875 -0.059615 -0.059615  1191000.0 -0.044889 -0.042614  1997      3   
3  121.5000  0.325153  0.325153  1191000.0  0.042396  0.058406  1997      4   
4  124.0000  0.020576  0.020576  1191000.0  0.071640  0.058577  1997      5   

   monthid  
0       13  
1       14  
2       15  
3       16  
4       17

In [10]:
compustat['YEAR'] = pd.DatetimeIndex(compustat['DATADATE']).year
compustat['MONTH'] = pd.DatetimeIndex(compustat['DATADATE']).month
compustat['monthid'] = ((compustat['YEAR']) - 1996) * 12 + compustat['MONTH']
sqldf("SELECT * FROM compustat LIMIT 5")

GVKEY                    DATADATE  FYEARQ  FQTR                    CONM  \
0  001161  1996-03-31 00:00:00.000000  1996.0   1.0  ADVANCED MICRO DEVICES   
1  001161  1996-06-30 00:00:00.000000  1996.0   2.0  ADVANCED MICRO DEVICES   
2  001161  1996-09-30 00:00:00.000000  1996.0   3.0  ADVANCED MICRO DEVICES   
3  001161  1996-12-31 00:00:00.000000  1996.0   4.0  ADVANCED MICRO DEVICES   
4  001161  1997-03-31 00:00:00.000000  1997.0   1.0  ADVANCED MICRO DEVICES   

   TIC DATACQTR DATAFQTR       ATQ      CEQQ  ...  CONSOL  INDFMT  DATAFMT  \
0  AMD   1996Q1   1996Q1  2954.809  2121.948  ...       C    INDL      STD   
1  AMD   1996Q2   1996Q2  2845.025  2071.355  ...       C    INDL      STD   
2  AMD   1996Q3   1996Q3  3053.992  2036.658  ...       C    INDL      STD   
3  AMD   1996Q4   1996Q4  3145.283  2021.878  ...       C    INDL      STD   
4  AMD   1997Q1   1997Q1  3426.126  2043.757  ...       C    INDL      STD   

  POPSRC CURCDQ COSTAT PERMNO  YEAR MONTH monthid  
0      D    USD      A  61241  1996     3       3  
1      D    USD      A  61241  1996     6       6  
2      D    USD      A  61241  1996     9       9  
3      D    USD      A  61241  1996    12      12  
4      D    USD      A  61241  1997     3      15  

[5 rows x 23 columns]

In [11]:
df_compustat_crsp = sqldf("SELECT a.*, b.* \
                           FROM crsp as a \
                           LEFT JOIN compustat as b \
                           ON a.PERMNO = b.PERMNO and a.monthid >= b.monthid + 4 and a.monthid <= b.monthid + 6") # 3-month lag

In [12]:
df_compustat_crsp

PERMNO                        DATE     CUSIP          COMNAM TICKER  \
0      10107.0  1997-01-31 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
1      10107.0  1997-02-28 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
2      10107.0  1997-03-31 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
3      10107.0  1997-04-30 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
4      10107.0  1997-05-30 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
...        ...                         ...       ...             ...    ...   
23998  93436.0  2022-08-31 00:00:00.000000  88160R10       TESLA INC   TSLA   
23999  93436.0  2022-09-30 00:00:00.000000  88160R10       TESLA INC   TSLA   
24000  93436.0  2022-10-31 00:00:00.000000  88160R10       TESLA INC   TSLA   
24001  93436.0  2022-11-30 00:00:00.000000  88160R10       TESLA INC   TSLA   
24002  93436.0  2022-12-30 00:00:00.000000  88160R10       TESLA INC   TSLA   

              PRC       RET      RETX     SHROUT    VWRETD  ...  CONSOL  \
0      102.000000  0.234493  0.234493  1198000.0  0.053473  ...       C   
1       97.500000 -0.044118 -0.044118  1198000.0 -0.001067  ...       C   
2       91.687500 -0.059615 -0.059615  1191000.0 -0.044889  ...       C   
3      121.500000  0.325153  0.325153  1191000.0  0.042396  ...       C   
4      124.000000  0.020576  0.020576  1191000.0  0.071640  ...       C   
...           ...       ...       ...        ...       ...  ...     ...   
23998  275.609985 -0.072489 -0.072489  3133470.0 -0.036233  ...       C   
23999  265.250000 -0.037589 -0.037589  3158000.0 -0.091323  ...       C   
24000  227.539993 -0.142168 -0.142168  3157752.0  0.077394  ...       C   
24001  194.699997 -0.144326 -0.144326  3157752.0  0.052354  ...       C   
24002  123.180000 -0.367334 -0.367334  3164000.0 -0.057119  ...       C   

       INDFMT  DATAFMT  POPSRC CURCDQ COSTAT  PERMNO    YEAR MONTH monthid  
0        INDL      STD       D    USD      A   10107  1996.0   9.0     9.0  
1        INDL      STD       D    USD      A   10107  1996.0   9.0     9.0  
2        INDL      STD       D    USD      A   10107  1996.0   9.0     9.0  
3        INDL      STD       D    USD      A   10107  1996.0  12.0    12.0  
4        INDL      STD       D    USD      A   10107  1996.0  12.0    12.0  
...       ...      ...     ...    ...    ...     ...     ...   ...     ...  
23998    INDL      STD       D    USD      A   93436  2022.0   3.0   315.0  
23999    INDL      STD       D    USD      A   93436  2022.0   3.0   315.0  
24000    INDL      STD       D    USD      A   93436  2022.0   6.0   318.0  
24001    INDL      STD       D    USD      A   93436  2022.0   6.0   318.0  
24002    INDL      STD       D    USD      A   93436  2022.0   6.0   318.0  

[24003 rows x 37 columns]

## Notice Duplicates
- Notice that our dataframe size exceeds our initial data size from CRSP of 23997 rows, so we have some duplicates
- And indeed we do (I found them using the command below, but showing them first so it's easier to understand)

In [13]:
df_compustat_crsp[6870:6881]

PERMNO                        DATE     CUSIP                  COMNAM  \
6870  45911.0  2002-04-30 00:00:00.000000  83088M10    ALPHA INDUSTRIES INC   
6871  45911.0  2002-05-31 00:00:00.000000  83088M10    ALPHA INDUSTRIES INC   
6872  45911.0  2002-05-31 00:00:00.000000  83088M10    ALPHA INDUSTRIES INC   
6873  45911.0  2002-06-28 00:00:00.000000  83088M10  SKYWORKS SOLUTIONS INC   
6874  45911.0  2002-06-28 00:00:00.000000  83088M10  SKYWORKS SOLUTIONS INC   
6875  45911.0  2002-07-31 00:00:00.000000  83088M10  SKYWORKS SOLUTIONS INC   
6876  45911.0  2002-07-31 00:00:00.000000  83088M10  SKYWORKS SOLUTIONS INC   
6877  45911.0  2002-08-30 00:00:00.000000  83088M10  SKYWORKS SOLUTIONS INC   
6878  45911.0  2002-08-30 00:00:00.000000  83088M10  SKYWORKS SOLUTIONS INC   
6879  45911.0  2002-09-30 00:00:00.000000  83088M10  SKYWORKS SOLUTIONS INC   
6880  45911.0  2002-09-30 00:00:00.000000  83088M10  SKYWORKS SOLUTIONS INC   

     TICKER    PRC       RET      RETX    SHROUT    VWRETD  ...  CONSOL  \
6870   AHAA  12.25 -0.196721 -0.196721   44260.0 -0.049600  ...       C   
6871   AHAA  10.19 -0.168163 -0.168163   44291.0 -0.010510  ...       C   
6872   AHAA  10.19 -0.168163 -0.168163   44291.0 -0.010510  ...       C   
6873   SWKS   5.55 -0.455348 -0.455348  137368.0 -0.070259  ...       C   
6874   SWKS   5.55 -0.455348 -0.455348  137368.0 -0.070259  ...       C   
6875   SWKS   2.90 -0.477477 -0.477477  137510.0 -0.081125  ...       C   
6876   SWKS   2.90 -0.477477 -0.477477  137510.0 -0.081125  ...       C   
6877   SWKS   4.20  0.448276  0.448276  137510.0  0.007949  ...       C   
6878   SWKS   4.20  0.448276  0.448276  137510.0  0.007949  ...       C   
6879   SWKS   4.53  0.078572  0.078572  137589.0 -0.099923  ...       C   
6880   SWKS   4.53  0.078572  0.078572  137589.0 -0.099923  ...       C   

      INDFMT  DATAFMT  POPSRC CURCDQ COSTAT  PERMNO    YEAR MONTH monthid  
6870    INDL      STD       D    USD      A   45911  2001.0  12.0    72.0  
6871    INDL      STD       D    USD      A   45911  2001.0  12.0    72.0  
6872    INDL      STD       D    USD      A   45911  2001.0  12.0    72.0  
6873    INDL      STD       D    USD      A   45911  2001.0  12.0    72.0  
6874    INDL      STD       D    USD      A   45911  2001.0  12.0    72.0  
6875    INDL      STD       D    USD      A   45911  2002.0   3.0    75.0  
6876    INDL      STD       D    USD      A   45911  2002.0   3.0    75.0  
6877    INDL      STD       D    USD      A   45911  2002.0   3.0    75.0  
6878    INDL      STD       D    USD      A   45911  2002.0   3.0    75.0  
6879    INDL      STD       D    USD      A   45911  2002.0   3.0    75.0  
6880    INDL      STD       D    USD      A   45911  2002.0   3.0    75.0  

[11 rows x 37 columns]

In [14]:
# These are our dropped rows
df_compustat_crsp[~df_compustat_crsp.index.isin(df_compustat_crsp.drop_duplicates(["PERMNO", "DATE"], ignore_index=False).index)]

PERMNO                        DATE     CUSIP                  COMNAM  \
6870  45911.0  2002-04-30 00:00:00.000000  83088M10    ALPHA INDUSTRIES INC   
6872  45911.0  2002-05-31 00:00:00.000000  83088M10    ALPHA INDUSTRIES INC   
6874  45911.0  2002-06-28 00:00:00.000000  83088M10  SKYWORKS SOLUTIONS INC   
6876  45911.0  2002-07-31 00:00:00.000000  83088M10  SKYWORKS SOLUTIONS INC   
6878  45911.0  2002-08-30 00:00:00.000000  83088M10  SKYWORKS SOLUTIONS INC   
6880  45911.0  2002-09-30 00:00:00.000000  83088M10  SKYWORKS SOLUTIONS INC   

     TICKER    PRC       RET      RETX    SHROUT    VWRETD  ...  CONSOL  \
6870   AHAA  12.25 -0.196721 -0.196721   44260.0 -0.049600  ...       C   
6872   AHAA  10.19 -0.168163 -0.168163   44291.0 -0.010510  ...       C   
6874   SWKS   5.55 -0.455348 -0.455348  137368.0 -0.070259  ...       C   
6876   SWKS   2.90 -0.477477 -0.477477  137510.0 -0.081125  ...       C   
6878   SWKS   4.20  0.448276  0.448276  137510.0  0.007949  ...       C   
6880   SWKS   4.53  0.078572  0.078572  137589.0 -0.099923  ...       C   

      INDFMT  DATAFMT  POPSRC CURCDQ COSTAT  PERMNO    YEAR MONTH monthid  
6870    INDL      STD       D    USD      A   45911  2001.0  12.0    72.0  
6872    INDL      STD       D    USD      A   45911  2001.0  12.0    72.0  
6874    INDL      STD       D    USD      A   45911  2001.0  12.0    72.0  
6876    INDL      STD       D    USD      A   45911  2002.0   3.0    75.0  
6878    INDL      STD       D    USD      A   45911  2002.0   3.0    75.0  
6880    INDL      STD       D    USD      A   45911  2002.0   3.0    75.0  

[6 rows x 37 columns]

In [15]:
# So we drop the duplicates and we are good
df_compustat_crsp.drop_duplicates(["PERMNO", "DATE"], inplace=True)
df_compustat_crsp

PERMNO                        DATE     CUSIP          COMNAM TICKER  \
0      10107.0  1997-01-31 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
1      10107.0  1997-02-28 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
2      10107.0  1997-03-31 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
3      10107.0  1997-04-30 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
4      10107.0  1997-05-30 00:00:00.000000  59491810  MICROSOFT CORP   MSFT   
...        ...                         ...       ...             ...    ...   
23998  93436.0  2022-08-31 00:00:00.000000  88160R10       TESLA INC   TSLA   
23999  93436.0  2022-09-30 00:00:00.000000  88160R10       TESLA INC   TSLA   
24000  93436.0  2022-10-31 00:00:00.000000  88160R10       TESLA INC   TSLA   
24001  93436.0  2022-11-30 00:00:00.000000  88160R10       TESLA INC   TSLA   
24002  93436.0  2022-12-30 00:00:00.000000  88160R10       TESLA INC   TSLA   

              PRC       RET      RETX     SHROUT    VWRETD  ...  CONSOL  \
0      102.000000  0.234493  0.234493  1198000.0  0.053473  ...       C   
1       97.500000 -0.044118 -0.044118  1198000.0 -0.001067  ...       C   
2       91.687500 -0.059615 -0.059615  1191000.0 -0.044889  ...       C   
3      121.500000  0.325153  0.325153  1191000.0  0.042396  ...       C   
4      124.000000  0.020576  0.020576  1191000.0  0.071640  ...       C   
...           ...       ...       ...        ...       ...  ...     ...   
23998  275.609985 -0.072489 -0.072489  3133470.0 -0.036233  ...       C   
23999  265.250000 -0.037589 -0.037589  3158000.0 -0.091323  ...       C   
24000  227.539993 -0.142168 -0.142168  3157752.0  0.077394  ...       C   
24001  194.699997 -0.144326 -0.144326  3157752.0  0.052354  ...       C   
24002  123.180000 -0.367334 -0.367334  3164000.0 -0.057119  ...       C   

       INDFMT  DATAFMT  POPSRC CURCDQ COSTAT  PERMNO    YEAR MONTH monthid  
0        INDL      STD       D    USD      A   10107  1996.0   9.0     9.0  
1        INDL      STD       D    USD      A   10107  1996.0   9.0     9.0  
2        INDL      STD       D    USD      A   10107  1996.0   9.0     9.0  
3        INDL      STD       D    USD      A   10107  1996.0  12.0    12.0  
4        INDL      STD       D    USD      A   10107  1996.0  12.0    12.0  
...       ...      ...     ...    ...    ...     ...     ...   ...     ...  
23998    INDL      STD       D    USD      A   93436  2022.0   3.0   315.0  
23999    INDL      STD       D    USD      A   93436  2022.0   3.0   315.0  
24000    INDL      STD       D    USD      A   93436  2022.0   6.0   318.0  
24001    INDL      STD       D    USD      A   93436  2022.0   6.0   318.0  
24002    INDL      STD       D    USD      A   93436  2022.0   6.0   318.0  

[23997 rows x 37 columns]

Finally, we keep the cleaned data from 2000-2022 for our next set of questions.

In [16]:
df_compustat_crsp_2000_2022 = df_compustat_crsp.loc[df_compustat_crsp["DATE"] >= "2000-01-01"]

In [17]:
df_compustat_crsp_2000_2022.reset_index(inplace=True)

In [18]:
df_compustat_crsp_2000_2022

index   PERMNO                        DATE     CUSIP          COMNAM  \
0         36  10107.0  2000-01-31 00:00:00.000000  59491810  MICROSOFT CORP   
1         37  10107.0  2000-02-29 00:00:00.000000  59491810  MICROSOFT CORP   
2         38  10107.0  2000-03-31 00:00:00.000000  59491810  MICROSOFT CORP   
3         39  10107.0  2000-04-28 00:00:00.000000  59491810  MICROSOFT CORP   
4         40  10107.0  2000-05-31 00:00:00.000000  59491810  MICROSOFT CORP   
...      ...      ...                         ...       ...             ...   
22015  23998  93436.0  2022-08-31 00:00:00.000000  88160R10       TESLA INC   
22016  23999  93436.0  2022-09-30 00:00:00.000000  88160R10       TESLA INC   
22017  24000  93436.0  2022-10-31 00:00:00.000000  88160R10       TESLA INC   
22018  24001  93436.0  2022-11-30 00:00:00.000000  88160R10       TESLA INC   
22019  24002  93436.0  2022-12-30 00:00:00.000000  88160R10       TESLA INC   

      TICKER         PRC       RET      RETX     SHROUT  ...  CONSOL  INDFMT  \
0       MSFT   97.875000 -0.161670 -0.161670  5160025.0  ...       C    INDL   
1       MSFT   89.375000 -0.086845 -0.086845  5160025.0  ...       C    INDL   
2       MSFT  106.250000  0.188811  0.188811  5242000.0  ...       C    INDL   
3       MSFT   69.750000 -0.343529 -0.343529  5262405.0  ...       C    INDL   
4       MSFT   62.562500 -0.103047 -0.103047  5262405.0  ...       C    INDL   
...      ...         ...       ...       ...        ...  ...     ...     ...   
22015   TSLA  275.609985 -0.072489 -0.072489  3133470.0  ...       C    INDL   
22016   TSLA  265.250000 -0.037589 -0.037589  3158000.0  ...       C    INDL   
22017   TSLA  227.539993 -0.142168 -0.142168  3157752.0  ...       C    INDL   
22018   TSLA  194.699997 -0.144326 -0.144326  3157752.0  ...       C    INDL   
22019   TSLA  123.180000 -0.367334 -0.367334  3164000.0  ...       C    INDL   

       DATAFMT  POPSRC  CURCDQ COSTAT PERMNO    YEAR  MONTH monthid  
0          STD       D     USD      A  10107  1999.0    9.0    45.0  
1          STD       D     USD      A  10107  1999.0    9.0    45.0  
2          STD       D     USD      A  10107  1999.0    9.0    45.0  
3          STD       D     USD      A  10107  1999.0   12.0    48.0  
4          STD       D     USD      A  10107  1999.0   12.0    48.0  
...        ...     ...     ...    ...    ...     ...    ...     ...  
22015      STD       D     USD      A  93436  2022.0    3.0   315.0  
22016      STD       D     USD      A  93436  2022.0    3.0   315.0  
22017      STD       D     USD      A  93436  2022.0    6.0   318.0  
22018      STD       D     USD      A  93436  2022.0    6.0   318.0  
22019      STD       D     USD      A  93436  2022.0    6.0   318.0  

[22020 rows x 38 columns]

***
## Task 4
We'll add the three additional variables `lnSize`, `bk2mkt`, `eP`

In [30]:
# a. lnSize
df_compustat_crsp_2000_2022.loc[:,"lnSize"] = np.log(df_compustat_crsp_2000_2022["CEQQ"])

# b. bk2mkt - assuming [ATQ] Assets – Total means ASSETS - LIABILITIES
df_compustat_crsp_2000_2022.loc[:,"bk2mkt"] = df_compustat_crsp_2000_2022["CEQQ"] / df_compustat_crsp_2000_2022["ATQ"]

# c. eP
df_compustat_crsp_2000_2022.loc[:,"eP1"] = df_compustat_crsp_2000_2022["IBQ"] / df_compustat_crsp_2000_2022["CEQQ"]
df_compustat_crsp_2000_2022.loc[:,"eP2"] = df_compustat_crsp_2000_2022["EPSPXQ"] / df_compustat_crsp_2000_2022["PRC"]

In [31]:
df_compustat_crsp_2000_2022

index   PERMNO                        DATE     CUSIP          COMNAM  \
0         36  10107.0  2000-01-31 00:00:00.000000  59491810  MICROSOFT CORP   
1         37  10107.0  2000-02-29 00:00:00.000000  59491810  MICROSOFT CORP   
2         38  10107.0  2000-03-31 00:00:00.000000  59491810  MICROSOFT CORP   
3         39  10107.0  2000-04-28 00:00:00.000000  59491810  MICROSOFT CORP   
4         40  10107.0  2000-05-31 00:00:00.000000  59491810  MICROSOFT CORP   
...      ...      ...                         ...       ...             ...   
22015  23998  93436.0  2022-08-31 00:00:00.000000  88160R10       TESLA INC   
22016  23999  93436.0  2022-09-30 00:00:00.000000  88160R10       TESLA INC   
22017  24000  93436.0  2022-10-31 00:00:00.000000  88160R10       TESLA INC   
22018  24001  93436.0  2022-11-30 00:00:00.000000  88160R10       TESLA INC   
22019  24002  93436.0  2022-12-30 00:00:00.000000  88160R10       TESLA INC   

      TICKER         PRC       RET      RETX     SHROUT  ...  CURCDQ  COSTAT  \
0       MSFT   97.875000 -0.161670 -0.161670  5160025.0  ...     USD       A   
1       MSFT   89.375000 -0.086845 -0.086845  5160025.0  ...     USD       A   
2       MSFT  106.250000  0.188811  0.188811  5242000.0  ...     USD       A   
3       MSFT   69.750000 -0.343529 -0.343529  5262405.0  ...     USD       A   
4       MSFT   62.562500 -0.103047 -0.103047  5262405.0  ...     USD       A   
...      ...         ...       ...       ...        ...  ...     ...     ...   
22015   TSLA  275.609985 -0.072489 -0.072489  3133470.0  ...     USD       A   
22016   TSLA  265.250000 -0.037589 -0.037589  3158000.0  ...     USD       A   
22017   TSLA  227.539993 -0.142168 -0.142168  3157752.0  ...     USD       A   
22018   TSLA  194.699997 -0.144326 -0.144326  3157752.0  ...     USD       A   
22019   TSLA  123.180000 -0.367334 -0.367334  3164000.0  ...     USD       A   

       PERMNO    YEAR  MONTH monthid     lnSize    bk2mkt       eP1       eP2  
0       10107  1999.0    9.0    45.0  10.320881  0.765275  0.072167  0.004393  
1       10107  1999.0    9.0    45.0  10.320881  0.765275  0.072167  0.004811  
2       10107  1999.0    9.0    45.0  10.320881  0.765275  0.072167  0.004047  
3       10107  1999.0   12.0    48.0  10.451291  0.767059  0.070427  0.006738  
4       10107  1999.0   12.0    48.0  10.451291  0.767059  0.070427  0.007512  
...       ...     ...    ...     ...        ...       ...       ...       ...  
22015   93436  2022.0    3.0   315.0  10.436613  0.516142  0.097345  0.011611  
22016   93436  2022.0    3.0   315.0  10.436613  0.516142  0.097345  0.012064  
22017   93436  2022.0    6.0   318.0  10.501664  0.530936  0.062101  0.009581  
22018   93436  2022.0    6.0   318.0  10.501664  0.530936  0.062101  0.011197  
22019   93436  2022.0    6.0   318.0  10.501664  0.530936  0.062101  0.017698  

[22020 rows x 42 columns]

In [35]:
# Now let's compare the standard deviation of the two methods of computing the earnings-to-price ratio

# eP1 is IBQ (Income) / Market Equity
print(f"eP1 - IBQ / Market Equity - mean: {df_compustat_crsp_2000_2022.eP1.mean():.4f} | stdev: {df_compustat_crsp_2000_2022.eP1.std():.4f}")

# eP2 is EPSPXQ (Earnings per share) / Prc
print(f"eP2 - EPSPXQ (Earnings per share) / Prc - mean: {df_compustat_crsp_2000_2022.eP2.mean():.4f} | stdev: {df_compustat_crsp_2000_2022.eP2.std():.4f}")


eP1 - IBQ / Market Equity - mean: -0.0370 | stdev: 3.8632
eP2 - EPSPXQ / Prc - mean: 0.0025 | stdev: 0.1539


## eP1 vs eP2

The two measures both illustrate how the market values the company's quarterly earnings. The difference is that `eP1`, which is calulated as total income / total market equity, represents the ratio at which the entire company is valued compared to its earnings. On the other hand, `eP2` represents how the company's earnings is valued at a per-share basis.

# TODO
Statistically, we may say that the latter (eP2) is the better comaprison because it has a smaller standard deviation 

## Analysis
- We will plot the distribution of each statistic, in an **aggregate** histogram including all stocks being analyzed.

## Question 5
- We wish to calculate some descriptive statistics for the variables we made above
- We again assume we are doing so by aggregating all of the stocks together.

In [92]:
statistics_df = pd.DataFrame(columns=["", "lnSize", "bk2mkt", "eP1", "eP2"])
statistics_df.loc[0] = "mean"
statistics_df.loc[1] = "median"
statistics_df.loc[2] = "variance"
statistics_df.loc[3] = "standard deviation"
statistics_df.loc[4] = "5th Percentile"
statistics_df.loc[5] = "25th Percentile"
statistics_df.loc[6] = "50th Percentile"
statistics_df.loc[7] = "75th Percentile"
statistics_df.loc[8] = "95th Percentile"

statistics_df.set_index("", inplace=True)

variables = ["lnSize", "bk2mkt", "eP1", "eP2"]


# Mean + Median
for var in variables:
    statistics_df.loc["mean", var] = df_compustat_crsp_2000_2022[var].mean()
    statistics_df.loc["median", var] = df_compustat_crsp_2000_2022[var].median()

# Variance + Std Deviation
for var in variables:
    statistics_df.loc["variance", var] = df_compustat_crsp_2000_2022[var].var()
    statistics_df.loc["standard deviation", var] = df_compustat_crsp_2000_2022[var].std()

# Percentiles
percentiles = [5, 25, 50, 75, 95]
for var in variables:
    for p in percentiles:
        statistics_df.loc[f"{p}th Percentile", var] = df_compustat_crsp_2000_2022[var].quantile(p / 100)

statistics_df.astype(float).round(4)

lnSize  bk2mkt      eP1     eP2
                                                    
mean                 8.0951  0.4942  -0.0370  0.0025
median               8.1394  0.5225   0.0406  0.0085
variance             2.8692  0.0781  14.9244  0.0237
standard deviation   1.6939  0.2795   3.8632  0.1539
5th Percentile       5.0846  0.0838  -0.1363 -0.0194
25th Percentile      7.0293  0.3197   0.0161  0.0033
50th Percentile      8.1394  0.5225   0.0406  0.0085
75th Percentile      9.2828  0.6997   0.0697  0.0132
95th Percentile     10.7642  0.8628   0.1582  0.0241

## Data Clean-Up

We will choose to **winsorize** our data, since the validity of our data should be good. I.e. the data that we have downloaded is not incorrect, but just reflects the notion that surprise stock events are a reality and should be reflected in the data. So, instead of truncating them to be removed, we just winsorize them so that they are included in the data, but we do not influence the skew of the data too much.

In [99]:
# Copy
df_compustat_crsp_2000_2022.loc[:,"lnSize (winsorized)"] = df_compustat_crsp_2000_2022.lnSize
df_compustat_crsp_2000_2022.loc[:,"bk2mkt (winsorized)"] = df_compustat_crsp_2000_2022.bk2mkt

df_compustat_crsp_2000_2022.loc[:,"eP1 (winsorized)"] = df_compustat_crsp_2000_2022.eP1
df_compustat_crsp_2000_2022.loc[:,"eP2 (winsorized)"] = df_compustat_crsp_2000_2022.eP2

# Winsorize
df_compustat_crsp_2000_2022.loc[df_compustat_crsp_2000_2022["lnSize (winsorized)"] > statistics_df.loc["mean", "lnSize"] + statistics_df.loc["standard deviation", "lnSize"] * 3] = statistics_df.loc["mean", "lnSize"] + statistics_df.loc["standard deviation", "lnSize"] * 3
df_compustat_crsp_2000_2022.loc[df_compustat_crsp_2000_2022["lnSize (winsorized)"] < statistics_df.loc["mean", "lnSize"] - statistics_df.loc["standard deviation", "lnSize"] * 3] = statistics_df.loc["mean", "lnSize"] - statistics_df.loc["standard deviation", "lnSize"] * 3

df_compustat_crsp_2000_2022.loc[df_compustat_crsp_2000_2022["bk2mkt (winsorized)"] > statistics_df.loc["mean", "bk2mkt"] + statistics_df.loc["standard deviation", "bk2mkt"] * 3] = statistics_df.loc["mean", "bk2mkt"] + statistics_df.loc["standard deviation", "bk2mkt"] * 3
df_compustat_crsp_2000_2022.loc[df_compustat_crsp_2000_2022["bk2mkt (winsorized)"] < statistics_df.loc["mean", "bk2mkt"] - statistics_df.loc["standard deviation", "bk2mkt"] * 3] = statistics_df.loc["mean", "bk2mkt"] - statistics_df.loc["standard deviation", "bk2mkt"] * 3

df_compustat_crsp_2000_2022.loc[df_compustat_crsp_2000_2022["eP1 (winsorized)"] > statistics_df.loc["mean", "eP1"] + statistics_df.loc["standard deviation", "eP1"] * 3] = statistics_df.loc["mean", "eP1"] + statistics_df.loc["standard deviation", "eP1"] * 3
df_compustat_crsp_2000_2022.loc[df_compustat_crsp_2000_2022["eP1 (winsorized)"] < statistics_df.loc["mean", "eP1"] - statistics_df.loc["standard deviation", "eP1"] * 3] = statistics_df.loc["mean", "eP1"] - statistics_df.loc["standard deviation", "eP1"] * 3

df_compustat_crsp_2000_2022.loc[df_compustat_crsp_2000_2022["eP2 (winsorized)"] > statistics_df.loc["mean", "eP2"] + statistics_df.loc["standard deviation", "eP2"] * 3] = statistics_df.loc["mean", "eP2"] + statistics_df.loc["standard deviation", "eP2"] * 3
df_compustat_crsp_2000_2022.loc[df_compustat_crsp_2000_2022["eP2 (winsorized)"] < statistics_df.loc["mean", "eP2"] - statistics_df.loc["standard deviation", "eP2"] * 3] = statistics_df.loc["mean", "eP2"] - statistics_df.loc["standard deviation", "eP2"] * 3

In [101]:
# Now we will redo the statistics df

winsorized_statistics_df = pd.DataFrame(columns=["", "lnSize (winsorized)", "bk2mkt (winsorized)", "eP1 (winsorized)", "eP2 (winsorized)"])
winsorized_statistics_df.loc[0] = "mean"
winsorized_statistics_df.loc[1] = "median"
winsorized_statistics_df.loc[2] = "variance"
winsorized_statistics_df.loc[3] = "standard deviation"
winsorized_statistics_df.loc[4] = "5th Percentile"
winsorized_statistics_df.loc[5] = "25th Percentile"
winsorized_statistics_df.loc[6] = "50th Percentile"
winsorized_statistics_df.loc[7] = "75th Percentile"
winsorized_statistics_df.loc[8] = "95th Percentile"

winsorized_statistics_df.set_index("", inplace=True)

variables = ["lnSize (winsorized)", "bk2mkt (winsorized)", "eP1 (winsorized)", "eP2 (winsorized)"]


# Mean + Median
for var in variables:
    winsorized_statistics_df.loc["mean", var] = df_compustat_crsp_2000_2022[var].mean()
    winsorized_statistics_df.loc["median", var] = df_compustat_crsp_2000_2022[var].median()

# Variance + Std Deviation
for var in variables:
    winsorized_statistics_df.loc["variance", var] = df_compustat_crsp_2000_2022[var].var()
    winsorized_statistics_df.loc["standard deviation", var] = df_compustat_crsp_2000_2022[var].std()

# Percentiles
percentiles = [5, 25, 50, 75, 95]
for var in variables:
    for p in percentiles:
        winsorized_statistics_df.loc[f"{p}th Percentile", var] = df_compustat_crsp_2000_2022[var].quantile(p / 100)

winsorized_statistics_df.astype(float).round(4)

lnSize (winsorized)  bk2mkt (winsorized)  \
                                                               
mean                             7.9714               0.4993   
median                           8.1032               0.5222   
variance                         3.8961               0.0652   
standard deviation               1.9739               0.2553   
5th Percentile                   4.7639               0.0856   
25th Percentile                  6.9771               0.3207   
50th Percentile                  8.1032               0.5222   
75th Percentile                  9.2726               0.6997   
95th Percentile                 10.7619               0.8627   

                    eP1 (winsorized)  eP2 (winsorized)  
                                                        
mean                          0.0282            0.0025  
median                        0.0404            0.0084  
variance                      0.1142            0.0030  
standard deviation            0.3379            0.0547  
5th Percentile               -0.1713           -0.0247  
25th Percentile               0.0153            0.0031  
50th Percentile               0.0404            0.0084  
75th Percentile               0.0695            0.0132  
95th Percentile               0.1585            0.0246

In [104]:
statistics_df.astype(float).round(4)

lnSize  bk2mkt      eP1     eP2
                                                    
mean                 8.0951  0.4942  -0.0370  0.0025
median               8.1394  0.5225   0.0406  0.0085
variance             2.8692  0.0781  14.9244  0.0237
standard deviation   1.6939  0.2795   3.8632  0.1539
5th Percentile       5.0846  0.0838  -0.1363 -0.0194
25th Percentile      7.0293  0.3197   0.0161  0.0033
50th Percentile      8.1394  0.5225   0.0406  0.0085
75th Percentile      9.2828  0.6997   0.0697  0.0132
95th Percentile     10.7642  0.8628   0.1582  0.0241